dmoj/cptbox/ptproc.cpp

```diff
@@ -103,6 +179,69 @@ int pt_process::protection_fault(int syscall) {
     return PTBOX_EXIT_PROTECTION;
 }
 
+#if PTBOX_PCRE
+bool pt_process::do_pcre_match(const char *path) {
+    int result;
+
+    result = pcre2_match(re_fs_read, (PCRE2_SPTR8) path, PCRE2_ZERO_TERMINATED,
+                         0, 0, re_fs_read_data, re_fs_read_context);
+    printf("Opening file: %s... %s\n", path, result >= 0 ? "allowed" : "denied");
+
+    if (result >= 0) {
+        return true;
+    } else if (result == PCRE2_ERROR_NOMATCH) {
+        return false;
+    } else {
+        fprintf(stderr, "PCRE2 match error: %d\n", result);
+        return false;
+    }
+}
+#endif
+
+bool pt_process::handle_open_call() {
+#if PTBOX_PCRE
+    if (!re_fs_read) {
+        return false;
+    }
+
+    int result;
+    unsigned long file_ptr = (unsigned long) debugger->arg0();
+    char *file = debugger->readstr(file_ptr, 4096);
```

Comment:
Unrelated to this patchset, but this is a function that has failure conditions and _should_ be able to return `nullptr`. But it doesn't; the underlying code will just segfault.